src/piton/jax.py

```diff
@@ -0,0 +1,1048 @@
+) -> Tuple[Array, None]:
+    """The backward pass for gather / scatter
+
+    The basic idea here is that we transpose the indices and apply it in reverse to get the backward pass.
+    """
+
+    a_dim, indices = res
+
+    indices_a = indices[:, 0]
+    indices_b = indices[:, 1]
+
+    sort_indices = jnp.argsort(indices_a)
+
+    # Note that these are transposed
+    trans_indices = jnp.stack((indices_b, indices_a), axis=-1)
+    trans_indices = trans_indices[sort_indices]
+
+    r = gather_scatter_p.bind(g, trans_indices, output_dim=a_dim)
+
+    return (r, None)
+
+
+gather_scatter.defvjp(gather_scatter_fwd, gather_scatter_bwd)
+
+# The actual primitive used to implement this operation
+gather_scatter_p = jax.core.Primitive("gather_scatter_p")
+
+
+def gather_scatter_abstract_eval(
+    a: Array, indices: Array, output_dim: int
+) -> jax.core.ShapedArray:
+    """The abstract shape computation for gather_scatter."""
+    # Our CUDA kernel currently assume that the inner size is a factor of 32.
+    # TODO: Remove this requirement
+    assert a.shape[1] % 32 == 0
+
+    # Our CUDA kernel is currently only implemented for the following types
+    # Note that this kernel is sorta incoherent for float16 anyways ...
+    assert a.dtype == jnp.float32
+    assert indices.dtype == jnp.uint32
+
+    return jax.core.ShapedArray(
+        (
+            output_dim,
+            a.shape[1],
+        ),
+        a.dtype,
+    )
+
+
+def gather_scatter_xla_translation(
+    ctx: jax.interpreters.xla.TranslationContext,
+    avals_in: Sequence[jax.core.AbstractValue],
+    avals_out: Sequence[jax.core.AbstractValue],
+    a: jax.interpreters.xla.XlaOp,
+    indices: jax.interpreters.xla.XlaOp,
+    *,
+    output_dim: int,
+) -> Sequence[jax.interpreters.xla.XlaOp]:
+    """Actually compute gather / scatter in hardware."""
+
+    if ctx.platform == "cpu":
+        # For our CPU implementation, we simply compile our fallback
+        computation = jax.xla_computation(
+            gather_scatter_fallback, static_argnums=(2,)
+        )(*avals_in, output_dim)
+
+        res = xla_client.ops.Call(ctx.builder, computation, [a, indices])
+        return [
+            xla_client.ops.GetTupleElement(res, i)
+            for i, _ in enumerate(avals_out)
+        ]
+
+    elif ctx.platform == "cuda":
+        # For our GPU implementation, we actually call out to CUDA code
+        (
+            input_shape,
+            indices_shape,
+        ) = avals_in
+
+        (result_shape,) = avals_out
+        assert isinstance(input_shape, ShapedArray)
+        assert isinstance(indices_shape, ShapedArray)
+        assert isinstance(result_shape, ShapedArray)
+
+        assert result_shape.shape[0] == output_dim
+        assert result_shape.shape[1] == input_shape.shape[1]
+
+        assert input_shape.dtype == result_shape.dtype
+
+        if input_shape.dtype == np.float32:
+            name = "float_gather_scatter"
```

Comment:
This is the C++ function call

Reply:
https://github.com/som-shahlab/piton/blob/jax/native/gather_scatter.cu#L142

Reply:
Maybe worth commenting this and also pointing that L26-L36 is where you import it and then register these functions. It was a bit ambiguous to me when I read the code how this name is being used by the CustomCallWithLayout.

Reply:
I switched from relative imports to absolute imports so it is more clear where all these functions are coming from.